internal/downlink/data/data.go

```diff
@@ -387,14 +391,177 @@ func setTXParameters(ctx *dataContext) error {
 	return nil
 }
 
-func setDataTXInfo(ctx *dataContext) error {
+func sfToSNR(sf int) float64 {
+	// Minimal SNR for SFs:
+	// SF 7:   -7.5
+	// SF 8:   -10
+	// SF 9:   -12.5
+	// SF 10:  -15
+	// SF 11:  -17.5
+	// SF 12:  -20
+	switch sf {
+	case 12:
+		return -20
+	case 11:
+		return -17.5
+	case 10:
+		return -15
+	case 9:
+		return -12.5
+	case 8:
+		return -10
+	case 7:
+		return -7.5
+	default:
+		return 0
+	}
+}
+
+func isRX2Only(ctx *dataContext) (b bool, err error) {
```

Comment:
With the above proposal, I think it would be good to split this function into multiple components, as currently it doing both the data-rate comparison and it calculates the link budget. Might be better to put the latter in a separate function.

Reply:
I totally agree, this one got quite long.